### ISSUE TITLE
XLA + Horovod + FP16

### ISSUE THREAD

**Issue by joeyearsley** [uber/horovod#145]
When running Horovod with XLA and `n` gpus, XLA will create `n` XLA instances on each GPU whilst only using the `i`th XLA instance on that card. Where `i` is the card number.

Have you ran into this/ Do you know a way around this? 

We have seen XLA dramatically speed up our code with NCCL replicated, however we are struggling to get Horovod up to the same speeds.
  
 Also, with FP16 it seems we can't use `float16` gradients in `hvd.allreduce()`, will this be added in the future or am I doing the interaction wrong between FP16 and Horovod?

**Comment by alsrgv** [comment]
Hi @joeyearsley,

Can you clarify a bit more how you use XLA, NCCL and Horovod?  In my mind, XLA is used for inference, so I want to understand the context a bit more.

Re. FP16 - are you planning to run pure FP16 training?  We don't have support for FP16 at the moment, but it's pretty easy to add.  I have tried FP16 training with TF benchmarks (see #119) and got 15-20% speedup, but that transmits gradients using FP32.

**Comment by joeyearsley** [comment]
Hi @alsrgv,

Sorry, I wasn't casting my logits to float32 so that fixed the nccl broadcasting in Horovod. 

So that now works, I see an improvement of:
- 167% (2.67x) using mixed precision and XLA _(vs no XLA)_
- 139% (2.39x) using FP32 and XLA _(vs no XLA)_
- 31% (1.31x) using mixed precision _(vs fp32)_ and XLA
- 20% (1.2x) using mixed precision  _(vs fp32)_ and no XLA

I use XLA via the [jit init](https://www.tensorflow.org/performance/xla/jit#turning_on_jit_compilation) (and [manual cpu xla def](https://www.tensorflow.org/performance/xla/jit#placing_operators_on_xla_devices)).
With just pure TF replicated mode I see an improvement of *2.3x* (FP32 Training) & *3x* (FP16 Training).

In comparison to Horovod though I can get an extra 75 imgs/second with XLA and no extra without XLA. 

*However*, these are preliminary experiments on our side - I'm still going through the graphs and XLA generated components to ensure the results are genuine. As they seem too good to be true.
Also, XLA affects every graph differently.

The original problem with XLA duplicating on instances seems to be more a TF problem so I'll open an issue there. 

If you happen to run an experiment with XLA enabled I would be interested to hear of any speed ups you see - I usually sample imgs/second after 1000 iterations as XLA takes some time to compile the subgraphs.

**Comment by alsrgv** [comment]
@joeyearsley, would be great if you could share your benchmarking code.

Replicated / vs Horovod with XLA speed-up you see may potentially be related to the fact that you're probably running one process in replicated set up, and N (# GPU) processes in Horovod case - which presents more work to XLA compiler as it independently needs to do N times more work.  What if you start your sampling in this case later, say after 4000 iterations?

Also, you can try to do mixed mode - use replication within a machine and Horovod across machines.

**Comment by joeyearsley** [comment]
@alsrgv Unfortunately our benchmarking code is currently not setup for synthetic data at the moment so we can't release it.

I'll try doing that, however we've realised now that with XLA our code doesn't converge - so we are currently digging into that, as the speed ups look great. 

I'll try that in the future, for now single servers do the job.
